Small Scientific Reproducibility, Config + MOM_parameter Refactor

manishvenu opened the pull request:
~Please see: https://github.com/CROCODILE-CESM/regional-mom6/pull/54~

[reproduced here the first post in https://github.com/CROCODILE-CESM/regional-mom6/pull/54 by @navidcy]


Closes https://github.com/COSIMA/regional-mom6/issues/198


### Changes/Explanation:

1. I was working with Gustavo on running NWA Boundary Conditions with regional-mom6. It seemed like the two things that mattered for reproducing boundary conditions (like GFDLs) was the regridding method (e.g. bilinear, source2d, etc..) and fill method (e.g. horizontal or vertical first, in what direction). I lightly added those arguments as 2 variables to the experiment class, the config files, and the regridding functions. 

    The `regridding_method` variable is a string that defaults to "bilinear", and the `fill_method` is a function that defaults to `regridding.fill_missing_data`.

2. Moved the config out of regional_mom6.py to a new script to take that logic out of the main class.

3. Moved all the MOM parameter editing stuff to new script as well
4. Adjust tests as needed, including splitting tides and parameters.

Goal: The main regional_mom6 script is cleaner and we split some of the complexity of regional_mom6 off into seperate simpler scripts

navidcy commented:
Thanks @manishvenu for this!
@angus-g would you be down to review this?

navidcy commented:
Clarification question:

> Goal: The main regional_mom6 script is cleaner and we split some of the complexity of regional_mom6 off into seperate simpler scripts

What do you mean by "regional_mom6 script" and "seperate simpler scripts"? What does "script" refers to? You referring to source code? Parts of notebook? User-generated Python `.py` scripts?

angus-g commented on regional_mom6/regional_mom6.py:
```diff
@@ -610,6 +501,8 @@ class experiment:
         create_empty (bool): If ``True``, the experiment object is initialized empty. This is used for testing and experienced user manipulation.
         expt_name (str): The name of the experiment (for config file use)
         boundaries (List[str]): List of (rectangular) boundaries to be set. Default is ``["south", "north", "west", "east"]``. The boundaries are set as (list index + 1) in MOM_override in the order of the list, and less than 4 boundaries can be set.
+        regridding_method (str): regridding method to use throughout the entire experiment. Default is ``'bilinear'``
```
Other options are?

manishvenu replied:
Done

angus-g commented on regional_mom6/regional_mom6.py:
```diff
@@ -610,6 +501,8 @@ class experiment:
         expt_name (str): The name of the experiment (for config file use)
         boundaries (List[str]): List of (rectangular) boundaries to be set. Default is ``["south", "north", "west", "east"]``. The boundaries are set as (list index + 1) in MOM_override in the order of the list, and less than 4 boundaries can be set.
+        regridding_method (str): regridding method to use throughout the entire experiment. Default is ``'bilinear'``
+        fill_method (Function): The fill function to be used after regridding datasets
```
Takes a dataset, returns a dataset? Defaults to backfilling missing data?

manishvenu replied:
Added more of a comment, thanks

angus-g commented on regional_mom6/regional_mom6.py:
```diff
@@ -1599,7 +1455,13 @@ def setup_ocean_state_boundaries(
             rotational_method (Optional[str]): Method to use for rotating the boundary velocities.
                 Default is ``EXPAND_GRID``.
+            regridding_method (Optional[str]): The type of regridding method to use. Defaults to self.regridding_method
+            fill_method (Function): Fill method to use throughout the function. Default is ``rgd.fill_missing_data``
```
I guess the default is _actually_ `self.fill_method`?

manishvenu replied:
Yes, thanks!

angus-g commented on regional_mom6/regional_mom6.py:
```diff
@@ -810,7 +713,7 @@ def __init__(
 
     def __str__(self) -> str:
-        return json.dumps(self.write_config_file(export=False, quiet=True), indent=4)
+        return json.dumps(Config.save_to_json(export=False, quiet=True), indent=4)
```
Do you need to pass `self` here? Also, `Config` doesn't have a `quiet` flag.

```suggestion
        return json.dumps(Config.save_to_json(self, export=False), indent=4)
```

manishvenu replied:
We do, thanks, I'll write a test case for this.

angus-g commented on regional_mom6/config.py:
```diff
@@ -0,0 +1,144 @@
+
+        Arguments:
+            obj (rm6.Experiment): The experiment object to save.
+            path (str): Path to write the config file to. If not provided, the file is written to the ``mom_run_dir`` directory.
```
If `path=None` (default) and `export=True` (default), it'll try to `open(None, "w")`

manishvenu replied:
Thanks!

angus-g commented on regional_mom6/config.py:
```diff
@@ -0,0 +1,144 @@
+                }
+
+        if export:
+            with open(path, "w") as f:
```
This is where there's no checking on the value of `path`.

manishvenu replied:
Yup, thanks

angus-g commented on regional_mom6/config.py:
```diff
@@ -0,0 +1,144 @@
+            Load experiment variables from a configuration file and generate the horizontal and vertical grids (``hgrid``/``vgrid``).
+        Computer-specific functionality eliminates the ability to pass file paths.
+
+        (This is basically another way to initialize and experiment.)
```
```suggestion
        (This is basically another way to initialize an experiment.)
```

manishvenu replied:
Added

angus-g commented on regional_mom6/config.py:
```diff
@@ -0,0 +1,144 @@
+    ):
+        """
+            Load experiment variables from a configuration file and generate the horizontal and vertical grids (``hgrid``/``vgrid``).
+        Computer-specific functionality eliminates the ability to pass file paths.
```
Not really sure what this is implying?

manishvenu replied:
Removing it!

angus-g commented on regional_mom6/utils.py:
```diff
@@ -384,3 +385,23 @@ def is_rectilinear_hgrid(hgrid: xr.Dataset, rtol: float = 1e-3) -> bool:
+        for path in paths:
+            all_files.extend(Path(path).glob(pattern))
+
+    if len(all_files) == 0:
+        if error_message is None:
+            return "No files found at the following paths: {} for the following patterns: {}".format(
+                paths, patterns
+            )
+        else:
+            return error_message
```
Seems like this should raise an exception rather than breaking type contract and returning a string?

manishvenu replied:
Yup! Thanks

manishvenu commented:
> Clarification question:
> 
> > Goal: The main regional_mom6 script is cleaner and we split some of the complexity of regional_mom6 off into seperate simpler scripts
> 
> What do you mean by "regional_mom6 script" and "seperate simpler scripts"? What does "script" refers to? You referring to source code? Parts of notebook? User-generated Python `.py` scripts?

If you look at the diff in the PR, there's two new python files in the package. This code used to be part of regional_mom6.py, and I moved it to seperate files in response to a comment by @angus-g (MOM parameter tools) and my own ideas. If you look at the bullet points in the first comment it expands on this a little bit.

navidcy commented:
@manishvenu there are some conflicts to resolve when you can.

navidcy commented on regional_mom6/MOM_parameter_tools.py:
```diff
@@ -0,0 +1,203 @@
+    directory, param_name, param_value=None, comment=None, override=True, delete=False
+):
+    """
+    **Requires MOM parameter files present in the run directory**
+
+    Changes a parameter in the ``MOM_input`` or ``MOM_override`` file. Returns original value, if there was one.
+    If `delete` keyword argument is `True` the parameter is removed. But note, that **only** parameters from
+    the ``MOM_override`` are be deleted since deleting parameters from ``MOM_input`` is not safe and can lead to errors.
+    If the parameter does not exist, it will be added to the file.
```
```suggestion
    Changes a parameter in the ``MOM_input`` or ``MOM_override`` file. Returns original value, if there was one.
    If `delete` keyword argument is `True` the parameter is removed. But note, that **only** parameters from
    the ``MOM_override`` are be deleted, since deleting parameters from ``MOM_input`` is not safe and can lead to errors.
    If the parameter does not exist, it will be added to the file.
    
    **MOM parameter files need to be present in the run directory**
```

navidcy replied:
> "...will be added to the file."

Which file? The`MOM_override`?

manishvenu replied:
Yup, added it

navidcy commented:
@angus-g could you have another look here? thanks!

navidcy commented on regional_mom6/regional_mom6.py:
```diff
@@ -14,18 +14,17 @@
-from collections import defaultdict
 import json
-import copy
+from regional_mom6 import MOM_parameter_tools as mpt
 from regional_mom6 import regridding as rgd
 from regional_mom6 import rotation as rot
+from regional_mom6.config import Config
```
are these new modules/submodules introduced by this PR?
do we need to include them here?
https://github.com/CROCODILE-CESM/regional-mom6/blob/minimal_scientific_reprod_cherrypick/docs/api.rst?plain=1

manishvenu replied:
Yeah, that's a good point, I'll add them to the autodocs

angus-g reviewed (commented):
Seems reasonable, thanks. I'll leave approval to @navidcy
